icicle/include/icicle/hash.h

```diff
@@ -0,0 +1,70 @@
+#pragma once
+
+#include "errors.h"
+#include "runtime.h"
+#include "icicle/utils/utils.h"
+#include "icicle/config_extension.h"
+
+#include <cstdint>
+#include <functional>
+
+typedef uint32_t limb_t;
+
+/*************************** Frontend & Backend shared APIs ***************************/
+namespace icicle {
+  /**
+   * @brief Configuration struct for the hash.
+   */
+  struct HashConfig {
+    bool are_inputs_on_device = false;  ///< True if inputs are on device, false if on host. Default is false.
+    bool are_outputs_on_device = false; ///< True if outputs are on device, false if on host. Default is false.
+    bool is_async = false; ///< True to run the hash asynchronously, false to run it synchronously. Default is false.
+    ConfigExtension* ext = nullptr; ///< Backend-specific extensions.
+  };
+
+  /**
+   * @brief Abstract class representing a hash function. In order to support structures like mmcs, we support
+   * side-inputs to the hash. These are inputs that are inserted in an intermediate stage of the calculations.
+   */
+  class Hash
+  {
+  public:
+    const int m_total_input_limbs;           // Total number of regular input limbs
```

Comment:
These names are a little confusing. In hashes it is usually called `rate`. So maybe smth like `input_rate` and `output_rate` would be better? Also, what does `m` stand for here?